allennlp/training/model.py

```diff
@@ -0,0 +1,89 @@
+from typing import Any, List, Dict, Union
+import torch
+
+
+class Model(torch.nn.Module):
+    """
+    This abstract class represents a model to be trained. Rather than relying on the
+    raw torch Module, we provide a slightly different API.
+    """
+
+    def forward(self,
```

Comment:
Motivations:
1) As we can build dynamic graphs, the loss function you compute may require some state from the model, which is most likely going to be generated in the `forward` function, so having the loss function created within the call to forward makes passing around this state very simple.
2) I really want to avoid having the Keras API of specifying losses and then computing them for the user, because it's so inflexible.

Instead, I guess the forward pass could set some state attribute, which then is used when calling `self.compute_loss` separately? This might be easier in the sense that your loss computation might not depend on much, so storing stuff here would be straightfoward. I guess the other thing we could have would be forward always returning a loss, which is just sometimes `None` to unify the return values?